OSD-7065: Conditional container push

2uasimojo opened the pull request:
To facilitate referencing token-refresher containers by digest rather than by tag, we need to address the "disappearing digest" problem (see the linked card for background).

This commit accomplishes this by:
- Adding a make target and shell script that wraps the existing `container-make` in a conditional, short-circuiting if the image already exists in the repository.
- Configuring Circle CI's `container-push` target to invoke the above rather than `container-make` directly.

TODO: We could be more efficient by also conditioning the `build` job.

https://issues.redhat.com/browse/OSD-7065

2uasimojo committed:
OSD-7065: Conditional container push

To facilitate referencing token-refresher containers by digest rather
than by tag, we need to address the "disappearing digest" problem (see
the linked card for background).

This commit accomplishes this by:
- Adding a make target and shell script that wraps the existing
`container-make` in a conditional, short-circuiting if the image already
exists in the repository.
- Configuring Circle CI's `container-push` target to invoke the above
rather than `container-make` directly.

TODO: We could be more efficient by also conditioning the `build` job.

OSD-7065

2uasimojo commented:
/test lint

squat commented on build/conditional-container-push.sh:
```diff
@@ -0,0 +1,82 @@
+    local image_uri=$1
+    local output
+
+    if output=$(skopeo inspect docker://"${image_uri}" 2>&1); then
```
We need some way to install the `skopeo` binary, no?

2uasimojo replied:
Oh, that's a fair point :P

Do you know of a way I can play with this `ubuntu-1604:201903-01` machine image somehow? [Installing skopeo](https://github.com/containers/skopeo/blob/master/install.md) is, for some reason, not a simple thing on ubuntu, so I would want a way to test out whatever I come up with.

(By the way, I would use `docker image inspect` instead, except that requires pulling the image locally, whereas skopeo works remote. Other options include making direct REST queries to the docker API on quay, e.g. via python.)

squat replied:
I always use `docker image inspect` as well. I don't think pulling the image down locally is a deal breaker. If it's simpler, let's go with that :+1:

2uasimojo replied:
Okay, did that. Put it in a separate commit, mostly so future-me can see that delta.

squat reviewed (changes requested):
Thanks for submitting this! Overall LGTM, I think there's one thing that needs your attention

2uasimojo committed:
Adapt image_exists_in_repo to use docker

...instead of skopeo; because we know docker exists in the CircleCI
`ubuntu-1604:201903-01` machine image, but skopeo doesn't, and is
[difficult to install](https://github.com/containers/skopeo/blob/master/install.md#ubuntu).

2uasimojo commented on build/conditional-container-push.sh:
```diff
@@ -0,0 +1,91 @@
+    if output=$(docker image pull "${image_uri}" 2>&1); then
+        # The image exists. Sanity check the output.
+        local report
+        if report=$(docker image inspect "${image_uri}" 2>&1); then
```
This needs to not glom together stdout and stderr. Recently (!) in jenkins, skopeo has started printing things to stderr when the command succeeds. See https://github.com/openshift/boilerplate/pull/162 for the fix.

2uasimojo replied:
Done via 3a601e0.

2uasimojo reviewed (commented):
/hold

2uasimojo committed:
Fix a bug in conditional container push

See https://github.com/openshift/boilerplate/pull/162

2uasimojo commented:
@squat bump

Looks like this is getting [attention](https://issues.redhat.com/browse/OSD-8032) from OSD for "egress locked environments".

bwplotka reviewed (approved):
Great, Thanks for nice comments on script. LGTM 💪🏽
